flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DefaultDispatcherRunner.java

```diff
@@ -89,11 +93,29 @@ public CompletableFuture<Void> closeAsync() {
             }
         }
 
+        Exception exception = null;
+        try {
+            leaderElectionService.stop();
+        } catch (Exception e) {
+            exception = e;
+        }
+        final Exception leaderElectionServiceShutdownException = exception;
+
         stopDispatcherLeaderProcess();
 
         FutureUtils.forward(previousDispatcherLeaderProcessTerminationFuture, terminationFuture);
 
-        return terminationFuture;
+        return terminationFuture.whenComplete(
+                (ignoredResult, terminationThrowable) -> {
+                    if (terminationThrowable != null) {
+                        throw new CompletionException(
+                                ExceptionUtils.firstOrSuppressed(
+                                        terminationThrowable,
+                                        leaderElectionServiceShutdownException));
```

Comment:
I kinda preferred how this was modeled as a completable future in the lifecycle manager.